### ISSUE TITLE
Jupyter notebook - specify C++ 14

### ISSUE THREAD

**Issue by erlanger** [vgvassilev/cling#115]
Is there a way to specify `--std=c++14` for a jupyter notebook?
I had to patch the source code before, but that doesn't seem to be a good solution. Maybe a configuration value in the kernel.json file?

I attach a sample C++14 notebook that can help in testing if you wish to use it.
[clingnotebookc++14.zip](https://github.com/vgvassilev/cling/files/532051/clingnotebookc.14.zip)

**Comment by vgvassilev** [comment]
I am not sure how to do it with notebooks but you can start cling in c++14 mode with `cling -std=c++14`. We have a meta issue about these problems at #112 and @SylvainCorlay could probably answer this i a better way.

**Comment by SylvainCorlay** [comment]
The jupyter kernel does not offer the option to pass command line options for now. Ideally, this would be done via traitlet-configurable attributes like for the regular ipython kernel.

Eventually, this would be fed into the command-line options passed here:

https://github.com/vgvassilev/cling/blob/master/tools/Jupyter/kernel/clingkernel.py#L94

If you want to hack this locally, you can change this into

```python 
strarr = ctypes.c_char_p * 5
argv = strarr(b"clingJupyter", b"-std=c++14", b"-I" + clingInstDir.encode('utf-8') + b"/include/",b"",b"")
```

**Comment by vgvassilev** [comment]
I think it would also make sense to extend the cling kernel in this direction. I'd be happy to review & integrate a PR there.

**Comment by SylvainCorlay** [comment]
There are multiple ways to do this:
 - make it a configuration attribute of the kernel.
 - expose multiple kernels for different flavors of the C++ language.

**Comment by vgvassilev** [comment]
If we opt for the latter we should be very careful these different flavors of C++ to not end up in the same translation unit (TU). In general, this is not recommended because you could end up with terrible to debug issues, or incompatible code. Eg. rtti vs nortti.

I'd opt for making it a config attribute and then if we needed we could think of ways to mix C++ in a safe way.

**Comment by SylvainCorlay** [comment]
I was not thinking of mixing c++ flavors. but just have multiple kernels, one for c++03, one for c++11 and another for 14 etc... The fact that all of them actually calls the same executable with different command line options is an implementation detail.

**Comment by vgvassilev** [comment]
I don't have a strong feeling about this. I am slightly leaning to the config attribute because it sounds simple and easier to maintain and incorporate future versions. Also, there are some things that are experimental only, eg. `-std=c++1z`. How would we signal this is experimental if we provide already a kernel for this.

**Comment by erlanger** [comment]
@SylvainCorlay idea of having different kernel "names" (e.g. c++14, c++11) is not opposite to having the config attribute, as I see it is only the way the configuration option is exposed to the end user. It doesn't really change the implementation much, except you offer two different kernel configurations, one running c++14 and one running c++11. I think --from the end user point of view-- it is much easier to choose a c++14 or c++11 kernel on the web front end that having to go back and change a configuration file.

This will also make it easier to exchange notebook files because you can just open the file with any other jupyter installation (the kernel is specified in the notebook file). Otherwise the user with whom you share the notebook will have to change configuration values and if he does not have access to the server will not be able to do it.
Hope this gives some more information from the user's point of view.

**Comment by dabro** [comment]
I agree with @erlanger.  Other languages expose multiple kernels, eg Scala 2.10 vs Scala 2.11, Python2 vs Python3.  These are then associated with a particular notebook and follow that file.

**Comment by erlanger** [comment]
Closing, it is fixed with #118
